Assets/Scripts/Networking/ConnectionHandler.cs

```diff
@@ -0,0 +1,134 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+using UnityEngine.UI;
+
+public class ConnectionHandler : Photon.PunBehaviour {
+
+	[SerializeField] string gameVersion = "0";
+
+	public List<Selectable> connectionReqdObject = new List<Selectable>();
+	List<bool> croStartingStates = new List<bool>();
+
+	bool isInRoom = false, startMatch = false;
+
+	public TextTweening infoText;
+
+	public List<string> waitingTexts = new List<string>()
+	{
+		"Waiting for an idle god",
+		".Waiting for an idle god.",
+		"..Waiting for an idle god..",
+		"...Waiting for an idle god..."
+	};
+
+	public string connectionSuccessText;
+
+	#region Unity Callbacks
+
+	private void Awake()
+	{
+		PhotonNetwork.autoJoinLobby = false;
+
+		PhotonNetwork.automaticallySyncScene = true;
+	}
+
+	private void Start()
+	{
+		PhotonNetwork.ConnectUsingSettings(gameVersion);
+
+		foreach (var item in connectionReqdObject)
+		{	
+			croStartingStates.Add(item.interactable);
+			item.interactable = false;
+		}
+	}
+
+	private void Update()
+	{
+		if (isInRoom)
+		{
+			if (PhotonNetwork.room.PlayerCount == 2 && !startMatch)
+			{
+				infoText.SetText(connectionSuccessText);
+
+				startMatch = true;
+
+				//Start game
+			}
+		}
+	}
+
+	#endregion
+
+
+	#region Primary Methods
+
+	/// <summary>
+	/// Start the connection process. 
+	/// - If already connected, we attempt joining a random room
+	/// - if not yet connected, Connect this application instance to Photon Cloud Network
+	/// </summary>
+	public void Connect(string ip)
+	{
+		// we check if we are connected or not, we join if we are , else we initiate the connection to the server.
+		if (PhotonNetwork.connected)
+		{
+			//Matchmaking
+			if (string.IsNullOrEmpty(ip))
+			{
+				// #Critical we need at this point to attempt joining a Random Room. If it fails, we'll get notified in OnPhotonRandomJoinFailed() and we'll create one.
+				PhotonNetwork.JoinRandomRoom();
+			}
+			else
+			{
+				//Stub. May add Direct Connect
+			}
+		}
+
+		else
+		{
+			// #Critical, we must first and foremost connect to Photon Online Server.
+			PhotonNetwork.ConnectUsingSettings(gameVersion);
+		}
+	}
+
+	#endregion
+
+
+	#region Photon.PunBehaviour CallBacks
+
+	public override void OnConnectedToPhoton()
+	{
+		infoText.SetText("Welcome");
+
+		isInRoom = startMatch = false;
+
+		//Set our UI interactability to its proper starting state
+		connectionReqdObject.ForEach(obj => obj.interactable = croStartingStates [connectionReqdObject.IndexOf(obj)]);
+	}
+
+	public override void OnDisconnectedFromPhoton()
+	{
+		infoText.SetText("Disconnected.");
+	}
+
+	public override void OnPhotonRandomJoinFailed(object [] codeAndMsg)
+	{
+		// #Critical: we failed to join a random room, maybe none exists or they are all full. No worries, we create a new room.
```

Comment:
Is it possible to fail to join a random room if, for instance, there's no network connection? Is there any value of `codeAndMsg` that may indicate the network is unavailable?

Reply:
Probably some value of `codeAndMsg` that does. This bit is copy and paste code.

Good news is we can assert that we have network connection immediately before this call. This is used when we're entering matchmaking (after connection to server), but there's no available room — so we create a new one.

Reply:
Fair enough - may want to check if there are any documented values of the parameter that prevent us from creating a room just in case